cpp/src/io/orc/writer_impl.cu

```diff
@@ -1121,96 +1147,163 @@ writer::impl::encoded_statistics writer::impl::gather_statistic_blobs(
     } else {
       desc->ts_scale = 0;
     }
+    col_stats_dtypes.push_back(desc->stats_dtype);
+    col_types.push_back(column.type());
     for (auto const& stripe : segmentation.stripes) {
-      auto& grp = stripe_stat_merge[column.index() * segmentation.num_stripes() + stripe.id];
-      grp.col   = stat_desc.device_ptr(column.index());
+      auto& grp       = stripe_stat_merge[column.index() * segmentation.num_stripes() + stripe.id];
+      grp.col_dtype   = column.type();
+      grp.stats_dtype = desc->stats_dtype;
       grp.start_chunk =
         static_cast<uint32_t>(column.index() * segmentation.num_rowgroups() + stripe.first);
       grp.num_chunks = stripe.size;
       for (auto rg_idx_it = stripe.cbegin(); rg_idx_it < stripe.cend(); ++rg_idx_it) {
```

Comment:
The pattern `!= end_iterator` is slightly better because it doesn't require a random access iterator and doesn't require a check for "before" ordering.
```suggestion
      for (auto rg_idx_it = stripe.cbegin(); rg_idx_it != stripe.cend(); ++rg_idx_it) {
```